File: web/src/app/core/run-stats.ts
// What a run costs: the LLM calls it made, counted by the pass that made
// them, and the projection shown before a run starts so the price is visible
// while it can still be changed.

import { ProviderDialect } from './chat-client';
import {
  ADEQUACY_MIN_BATCHES,
  ADEQUACY_SAMPLE_FRACTION,
  ESTIMATED_SECS_PER_CALL,
  FIX_FLAGGED_FRACTION,
  FIX_FLAGGED_MIN,
  SYSTEMATIC_REPAIR_FRACTION,
} from './constants';

export const CALL_KINDS = [
  'scan', 'attribution', 'translate', 'review', 'repair', 'backTranslation',
] as const;

export type CallKind = (typeof CALL_KINDS)[number];
export type CallCounts = Record<CallKind, number>;

export const CALL_KIND_LABELS: Record<CallKind, string> = {
  scan: 'scan',
  attribution: 'attribution',
  translate: 'translate',
  review: 'review',
  repair: 'repair',
  backTranslation: 'back-translation',
};

export function emptyCallCounts(): CallCounts {
  return {
    scan: 0, attribution: 0, translate: 0, review: 0, repair: 0,
    backTranslation: 0,
  };
}

export function addCallCounts(a: CallCounts, b: CallCounts): CallCounts {
  const sum = emptyCallCounts();
  for (const kind of CALL_KINDS) sum[kind] = a[kind] + b[kind];
  return sum;
}

export function totalCalls(counts: CallCounts): number {
  return CALL_KINDS.reduce((sum, kind) => sum + counts[kind], 0);
}

/** "1 scan, 38 translate, 2 repair" — passes that made no call are left out. */
export function describeCalls(counts: CallCounts): string {
  return CALL_KINDS
    .filter((kind) => counts[kind] > 0)
    .map((kind) => `${counts[kind]} ${CALL_KIND_LABELS[kind]}`)
    .join(', ');
}

/** What one finished file cost. */
export interface RunStats {
  calls: CallCounts;
  blocks: number;
  elapsedMs: number;
  /** The request shape the provider turned out to need. */
  dialect?: ProviderDialect;
}

export function blocksPerSecond(blocks: number, elapsedMs: number): number {
  return elapsedMs > 0 ? (blocks * 1000) / elapsedMs : 0;
}

// Rounding up, never to nearest: the two trees must agree, and Python rounds
// halves to even where JavaScript rounds them up.
/** How many batches the flagged-batch repair may re-issue. */
export function fixFlaggedCap(batchCount: number): number {
  return Math.max(FIX_FLAGGED_MIN, Math.ceil(batchCount * FIX_FLAGGED_FRACTION));
}

/** The wider ceiling for a file where one cause keeps recurring: the ordinary
 * cap answers a handful of one-offs, and leaves a systematic failure almost
 * entirely unrepaired. */
export function systematicRepairCap(batchCount: number): number {
  return Math.max(
    FIX_FLAGGED_MIN, Math.ceil(batchCount * SYSTEMATIC_REPAIR_FRACTION));
}

/** How many batches the adequacy check back-translates. */
export function adequacySampleSize(batchCount: number): number {
  if (batchCount <= 0) return 0;
  return Math.min(
    batchCount,
    Math.max(ADEQUACY_MIN_BATCHES, Math.ceil(batchCount * ADEQUACY_SAMPLE_FRACTION)),
  );
}

/** Every Nth batch, so the sample spans the file rather than its opening. */
export function sampleIndices(batchCount: number, wanted: number): number[] {
  if (batchCount <= 0 || wanted <= 0) return [];
  const take = Math.min(batchCount, wanted);
  const step = batchCount / take;
  const picked: number[] = [];
  for (let i = 0; i < take; i++) picked.push(Math.floor(i * step));
  return picked;
}

export interface ProjectionInput {
  /** Blocks per queued file; each file is batched on its own. */
  blockCounts: number[];
  batchSize: number;
  /** Requests in flight at once: concurrency x parallel files. */
  lanes: number;
  review: boolean;
  fixFlagged: boolean;
  verifyAdequacy: boolean;
}

export interface RunProjection {
  calls: CallCounts;
  total: number;
  estimateMs: number;
  secsPerCall: number;
}

/** The call breakdown a run would make, and how long it would take. Speaker
 * attribution is absent on purpose: how many scenes qualify is only known once
 * the scan has replied, so counting it here would be a guess dressed as a
 * number. Review is an upper bound; repair is projected at the ordinary cap,
 * since the wider systematic ceiling depends on what the run turns out to
 * flag — quoting it for every file would overstate the ordinary one fivefold. */
export function projectRun(input: ProjectionInput): RunProjection {
  const calls = emptyCallCounts();
  const batchSize = Math.max(1, input.batchSize);
  for (const blocks of input.blockCounts) {
    if (blocks <= 0) continue;
    const batches = Math.ceil(blocks / batchSize);
    calls.scan += 1;
    calls.translate += batches;
    if (input.review) calls.review += batches;
    if (input.fixFlagged) calls.repair += fixFlaggedCap(batches);
    if (input.verifyAdequacy) calls.backTranslation += adequacySampleSize(batches);
  }
  const total = totalCalls(calls);
  const lanes = Math.max(1, input.lanes);
  return {
    calls,
    total,
    estimateMs: (total / lanes) * ESTIMATED_SECS_PER_CALL * 1000,
    secsPerCall: ESTIMATED_SECS_PER_CALL,
  };
}

File: cli/core/run_stats.py
"""What a run costs: the LLM calls it made, counted by the pass that made
them, and the projection --dry-run shows before a run starts, while the price
can still be changed."""

from __future__ import annotations

import math
from collections.abc import Callable, Mapping, Sequence
from dataclasses import dataclass, field

from .constants import (
    ADEQUACY_MIN_BATCHES,
    ADEQUACY_SAMPLE_FRACTION,
    FIX_FLAGGED_FRACTION,
    FIX_FLAGGED_MIN,
    SYSTEMATIC_CAUSE_MIN,
    SYSTEMATIC_REPAIR_FRACTION,
)

# Ordered as a run makes them; reports follow this order.
CALL_KINDS = ("scan", "attribution", "translate", "review", "repair",
              "back_translation")

CALL_KIND_LABELS = {
    "scan": "scan",
    "attribution": "attribution",
    "translate": "translate",
    "review": "review",
    "repair": "repair",
    "back_translation": "back-translation",
}


@dataclass
class CallCounts:
    """LLM calls made so far, by pass. Mutated in place by the passes
    themselves, so one run's config carries one running total."""
    scan: int = 0
    attribution: int = 0
    translate: int = 0
    review: int = 0
    repair: int = 0
    back_translation: int = 0
    # Told about every call as it goes out; the status line's meter listens.
    listener: Callable[[str], None] | None = field(default=None, compare=False, repr=False)

    def count(self, kind: str, times: int = 1) -> None:
        setattr(self, kind, getattr(self, kind) + times)
        if self.listener is not None:
            for _ in range(times):
                self.listener(kind)

    @property
    def total(self) -> int:
        return sum(getattr(self, k) for k in CALL_KINDS)

    def snapshot(self) -> CallCounts:
        return CallCounts(**{kind: getattr(self, kind) for kind in CALL_KINDS})

    def since(self, before: CallCounts) -> CallCounts:
        """What has been counted since `before` was taken."""
        return CallCounts(**{
            kind: getattr(self, kind) - getattr(before, kind) for kind in CALL_KINDS
        })


def describe_calls(counts: CallCounts) -> str:
    """"1 scan, 38 translate, 2 repair" — passes that made no call are left out."""
    return ", ".join(
        f"{getattr(counts, kind)} {CALL_KIND_LABELS[kind]}"
        for kind in CALL_KINDS if getattr(counts, kind)
    )


# Rounding up, never to nearest: the two trees must agree, and Python rounds
# halves to even where JavaScript rounds them up.
def fix_flagged_cap(batch_count: int) -> int:
    """How many batches the flagged-batch repair may re-issue when every cause
    is a one-off."""
    return max(FIX_FLAGGED_MIN, math.ceil(batch_count * FIX_FLAGGED_FRACTION))


def systematic_repair_cap(batch_count: int) -> int:
    """The wider ceiling for a file where one cause keeps recurring: the
    ordinary cap answers a handful of one-offs, and leaves a systematic failure
    almost entirely unrepaired."""
    return max(FIX_FLAGGED_MIN,
               math.ceil(batch_count * SYSTEMATIC_REPAIR_FRACTION))


def group_by_cause(
    flagged: Mapping[int, Sequence[str]],
) -> dict[str, list[int]]:
    """Cause -> the flagged batches carrying it, each in block order."""
    groups: dict[str, list[int]] = {}
    for index in sorted(flagged):
        for cause in dict.fromkeys(flagged[index]):
            groups.setdefault(cause, []).append(index)
    return groups


@dataclass(frozen=True)
class RepairPlan:
    """Which flagged batches to re-issue, and the arithmetic that chose them."""
    selected: list[int]
    cap: int
    causes: int
    flagged: int

    @property
    def skipped(self) -> int:
        return self.flagged - len(self.selected)


def plan_repairs(
    flagged: Mapping[int, Sequence[str]], batch_count: int,
) -> RepairPlan:
    """Pick the batches to re-issue by cause rather than by position.

    Rarest cause first, so one term drifting across the file cannot spend the
    whole budget and crowd out the batches nothing else will look at; within a
    cause, earliest batch first. The ceiling rises only when some cause has
    gone systematic — a run of unrelated one-offs is not evidence of a problem
    worth five times the calls.
    """
    groups = group_by_cause(flagged)
    systematic = any(len(batches) >= SYSTEMATIC_CAUSE_MIN
                     for batches in groups.values())
    cap = (systematic_repair_cap(batch_count) if systematic
           else fix_flagged_cap(batch_count))

    selected: list[int] = []
    seen: set[int] = set()
    # A stable sort, so causes of equal size keep the order they were first
    # flagged in rather than an alphabetical one.
    for cause in sorted(groups, key=lambda c: len(groups[c])):
        for index in groups[cause]:
            if len(selected) >= cap:
                break
            if index not in seen:
                seen.add(index)
                selected.append(index)
    return RepairPlan(sorted(selected), cap, len(groups), len(flagged))


def adequacy_sample_size(batch_count: int) -> int:
    """How many batches the adequacy check back-translates."""
    if batch_count <= 0:
        return 0
    return min(batch_count,
               max(ADEQUACY_MIN_BATCHES,
                   math.ceil(batch_count * ADEQUACY_SAMPLE_FRACTION)))


def sample_indices(batch_count: int, wanted: int) -> list[int]:
    """Every Nth batch, so the sample spans the file rather than its opening."""
    if batch_count <= 0 or wanted <= 0:
        return []
    take = min(batch_count, wanted)
    step = batch_count / take
    return [int(i * step) for i in range(take)]
